### PR TITLE
kernel/environ: Fix an error in the duplication of the child tasks environment

### PR DESCRIPTION
**Pull request by lokeshbv**
Fix an error in the duplication of the child tasks environment in the special case
where the parent's environment was created, but then all of the variables were unset.
In that case, there is still an allocation in place but the size of the allocation is zero.
This case was not being handled correctly when a child task attempts to create its environment
and inherit the zero-size parent environment. Noted by Anthony Merlino.

`This patch is back-ported from Nuttx.`

### CODE REVIEW

**Comment by lokeshbv** [comment]
But when `envlen = 0`,  it should not return `-ENOMEM`. `envlen = 0` means that zero size parent's environment, which is not an error.

File: os/kernel/environ/env_dup.c [hunk 1]
```diff
@@ -113,18 +113,39 @@ int env_dup(FAR struct task_group_s *group)
...
+			} else {
+				/* Duplicate the parent environment. */
+
+				memcpy(envp, ptcb->group->tg_envp, envlen);
+			}
 		}
+
+		/* Save the size and child environment allocation. */
+
+		group->tg_envsize = envlen;
```
[the remark this answers is not in the document]
Comment (lokeshbv) [reply os/kernel/environ/env_dup.c]: 1. yes, it is redundant when `envlen = 0`, but this is for the case when `envlen != 0`.
2. I don't get why no parent env is a problem. Can you please elaborate the problem.

[the remark this answers is not in the document]
Comment (sunghan-chang) [reply os/kernel/environ/env_dup.c]: 1. No. This line will be executed even if envlen is zero.
2. This function is duplication. Some logic calls this function to duplicate parent env. But regardless of reason, duplication is failed. If it is not a problem, there is no reason to be ENOENT or ENOANO or ENXIO these kinds errno.

**Comment by seinfra** [comment]
Target : [86ca89f52f7cea19a07cabc0304fca99538a43a7] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [86ca89f52f7cea19a07cabc0304fca99538a43a7] - Code Rule Check OK.

File: os/kernel/environ/env_dup.c [hunk 2]
```diff
@@ -113,17 +113,38 @@ int env_dup(FAR struct task_group_s *group)
...
+		if (envlen > 0) {
+			/* There is an environment, duplicate it */
+
+			envp = (FAR char *)kumm_malloc(envlen);
+			if (envp == NULL) {
+				/* The parent's environment can not be inherited due to a
+				 * failure in the allocation of the child environment.
+				 */
+
+				envlen = 0;
```
Comment (sunghan-chang) [inline os/kernel/environ/env_dup.c]: same as previous

File: os/kernel/environ/env_dup.c [hunk 3]
```diff
@@ -113,17 +113,38 @@ int env_dup(FAR struct task_group_s *group)
...
 
 		envlen = ptcb->group->tg_envsize;
-		envp   = (FAR char *)kumm_malloc(envlen);
-		if (!envp) {
-			ret = -ENOMEM;
-		} else {
-			group->tg_envsize = envlen;
-			group->tg_envp = envp;
-			memcpy(envp, ptcb->group->tg_envp, envlen);
+		envp   = NULL;
```
Comment (sunghan-chang) [inline os/kernel/environ/env_dup.c]: redundant. This line is necessary on Nuttx change. Not in here.

Comment (lokeshbv) [reply os/kernel/environ/env_dup.c]: removed

**Comment by seinfra** [comment]
Target : [87b1a941cf16faefc3de4f607533d04958266679] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [87b1a941cf16faefc3de4f607533d04958266679] - Code Rule Check OK.

**Pull request merged**